HPCC-29609 Make vcpkg-catalog file name correspond to CPack package name

timothyklemm opened the pull request:
Change vcpkg.cmake to create the CPack package file name. All top level projects that include this file gain access to a standardized name.
- Include the platform's version.cmake to ensure default version numbering is available.
- Include the top level project's version.cmake enabling projects to override default version numbering.
- Include the top level project's optional prePackageSetup.cmake enabling projects to configure any other default naming values.
- Include the platform's packageSetup.cmake to generate the package file name.
- Include the top level project's optional postPackagSetup.cmake isolating project-specific logic currently coupled with name generating logic.
- Allow top level projects to disregard platform naming conventions.

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

Commit:
HPCC-29609 Make vcpkg-catalog file name correspond to CPack package name

Change vcpkg.cmake to create the CPack package file name. All top level
projects that include this file gain access to a standardized name.
- Include the platform's version.cmake to ensure default version numbering
  is available.
- Include the top level project's version.cmake enabling projects to override
  default version numbering.
- Include the top level project's optional prePackageSetup.cmake enabling
  projects to configure any other default naming values.
- Include the platform's packageSetup.cmake to generate the package file name.
- Include the top level project's optional postPackagSetup.cmake isolating
  project-specific logic currently coupled with name generating logic.
- Allow top level projects to disregard platform naming conventions.

Signed-off-by: Tim Klemm <tim.klemm@lexisnexisrisk.com>

GordonSmith reviewed (approved):
Looks clean - when its imported where in the CMakeLists.txt file would it go (ideally not before the primary "project" statement?

timothyklemm commented on cmake_modules/packageSetup.cmake:
```diff
@@ -0,0 +1,246 @@
+        endif()
+    endif()
+    if ("${PACKAGE_CONTACT}" STREQUAL "")
+        set(PACKAGE_CONTACT "HPCCSystems <ossdevelopment@lexisnexis.com>")
```
I don't know exactly what CPack does with it, but we assign this value to CPACK_PACKAGE_CONTACT and CPACK_NSIS_CONTACT, and CPack itself defaults CPACK_DEBIAN_PACKAGE_CONTACT to this value.

GordonSmith replied:
I meant where it goes if someone emails to it!

timothyklemm commented:
I didn't spend time trying to identify the cause, but cmake failed every time I tried to include vcpkg.cmake _after_ the project statement. It seems like everything here, and in the corresponding LN PR (still to come), has to be before the project statement.

timothyklemm commented:
@GordonSmith, should this target candidate-9.2.x instead of master?

GordonSmith commented:
Yes - I would like to investigate if it can be refactored to work after the "project".

Also have you tested with LN?

timothyklemm commented:
@GordonSmith this newest commit simplifies pre-project steps in a way that can be used by other projects to decouple themselves from vcpkg compliance requirements. Also, by moving the name generation after the project statement, the LN project requires no changes.

GordonSmith reviewed (changes requested):
With the addition of the packageSetup - I was expecting more to be removed from the main CMakeLists.txt file?

timothyklemm commented on cmake_modules/packageSetup.cmake:
```diff
@@ -0,0 +1,285 @@
+        elseif (PLATFORM)
+            set(PACKAGE_FILE_NAME_PREFIX "hpccsystems-platform-${PACKAGE_PROJECT}")
+            set(CPACK_PACKAGE_NAME "${PACKAGE_FILE_NAME_PREFIX}")
+        else ()
```
Maybe, but I was trying to keep the existing logic as much as possible. If I make this an elseif what should be the result if the test fails?

Michael-Gardner replied:
This should be fine, it's how we have it set currently.

timothyklemm commented on CMakeLists.txt:
```diff
@@ -84,10 +79,30 @@ project(hpccsystems-platform LANGUAGES C CXX
-if(NOT CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
-    set(TOP_LEVEL_PROJECT OFF)
-endif(NOT CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
+set(TOP_LEVEL_PROJECT PROJECT_IS_TOP_LEVEL)
```
It can be removed.

timothyklemm commented:
> With the addition of the packageSetup - I was expecting more to be removed from the main CMakeLists.txt file?

I have been able to remove the redundant logic for platform builds, as well as my security manager package. The LN package fails in CPack, and I haven't identified the issue.

timothyklemm commented:
@GordonSmith this addresses your comments except for a possible CLIENT_TOOLS_ONLY check. I've tested builds and installs of the platform, LN, and my plugin. https://github.com/hpcc-systems/LN/pull/1026 includes the LN changes, which are now dependent on this PR.

timothyklemm commented:
@GordonSmith updated and rebased to master...

GordonSmith commented:
@Michael-Gardner - can you test this with a view to getting it merged to master prior 9.10.x?

timothyklemm commented:
@Michael-Gardner - reminder that https://github.com/hpcc-systems/LN/pull/1026 is a companion PR in LN that should target the same build version.

Michael-Gardner reviewed (changes requested):
Tim when I go to build this locally just to do a quick sanity check, I'm getting build errors. For some reason it says that your branch is way out of sync with master even though it doesn't look like it from the PR's commit page.

Can you squash your commits and make sure it has a fresh rebase w master? The code looks fine and I haven't seen any issues. I just want to do some local builds to ensure that everything still is functioning properly after these large changes.
